Add v8 coverage

M4tiz opened the pull request.

wouterlucas reviewed (approved):
👍

frank-weindel commented:
Investigating this. This change is somehow producing errors during the TypeScript build:
```
> @lightningjs/renderer@0.9.2 build /work
> tsc --build

node_modules/.pnpm/vite@4.5.3_@types+node@18.19.36/node_modules/vite/dist/node/index.d.ts:843:22 - error TS2420: Class 'import("/work/node_modules/.pnpm/vite@4.5.3_@types+node@18.19.36/node_modules/vite/dist/node/index").FSWatcher' incorrectly implements interface 'import("fs").FSWatcher'.
  Type 'FSWatcher' is missing the following properties from type 'FSWatcher': ref, unref

843 export declare class FSWatcher extends EventEmitter implements fs.FSWatcher {
                         ~~~~~~~~~

src/core/text-rendering/font-face-types/TrFontFace.ts:185:7 - error TS2353: Object literal may only specify known properties, and 'variant' does not exist in type 'FontFaceDescriptors'.

185       variant: descriptors.variant,
          ~~~~~~~

src/core/text-rendering/font-face-types/WebTrFontFace.ts:60:7 - error TS2353: Object literal may only specify known properties, and 'variant' does not exist in type 'FontFaceDescriptors'.

60       variant: determinedDescriptors.variant,
         ~~~~~~~

node_modules/.pnpm/vite@4.5.3_@types+node@18.19.36/node_modules/vite/dist/node/index.d.ts:843:22 - error TS2420: Class 'import("/work/node_modules/.pnpm/vite@4.5.3_@types+node@18.19.36/node_modules/vite/dist/node/index").FSWatcher' incorrectly implements interface 'import("fs").FSWatcher'.
  Type 'FSWatcher' is missing the following properties from type 'FSWatcher': ref, unref

843 export declare class FSWatcher extends EventEmitter implements fs.FSWatcher {
                         ~~~~~~~~~

src/core/text-rendering/font-face-types/TrFontFace.ts:185:7 - error TS2353: Object literal may only specify known properties, and 'variant' does not exist in type 'FontFaceDescriptors'.

185       variant: descriptors.variant,
          ~~~~~~~

src/core/text-rendering/font-face-types/WebTrFontFace.ts:60:7 - error TS2353: Object literal may only specify known properties, and 'variant' does not exist in type 'FontFaceDescriptors'.

60       variant: determinedDescriptors.variant,
         ~~~~~~~


Found 6 errors.
```

frank-weindel commented:
Okay. This issue is reproducible locally if you delete all the `node_modules` folders and run `pnpm install` from the repo root.

What seems to be happening is that the new dependency pulling in vite@4.5.3 (likely somehow via peerDependency), whereas prior to this change vite@4.5.0 would get installed.

Vite 4.5.2 seemed to have introduced this bug which seems to be fixed in Vite 5, but never backported to Vite 4:
- https://github.com/vitejs/vite/issues/15714#issuecomment-1933724182

Upgrading to Vite 5 may solve this issue. If you could help try that, that would be greatly appreciated and we could get this PR merged sooner. Thanks 🙏

wouterlucas commented:
Nice! I'd rebase this against `dev` - thats the working branch we're on and will be part of the next big "release". 

That way we don't even have to bother with stuff like:
```
[dev] 13 |      try {
[dev] 14 |          console.log('Loading core extension', coreExtensionModule);
[dev] 15 |          module = (await import(coreExtensionModule /* @vite-ignore */));
[dev]    |                                 ^^^^^^^^^^^^^^^^^^^
[dev] 16 |      }
[dev] 17 |      catch (e) {
```

😄

M4tiz commented:
@frank-weindel @wouterlucas 
I decided to try to update to Vite 5.x I hope that's alright :)

Crucial changes:
- Removed `variant` field since the new `FontFaceDescriptors` contract doesn't have it anymore. It seems to be compliant with https://www.w3.org/TR/css-fonts-3/#font-face-rule specification
- Changed `tsconfig` module resolution to `Node16` in all variants due to https://github.com/rollup/rollup/issues/5199

Let me know any of these require some additional checks to look for regression.

frank-weindel commented on src/core/utils.ts:
```diff
@@ -204,9 +204,9 @@ export const getTimingFunction = (
   }
 
   if (timingLookup[str] !== undefined) {
+    const [a, b, c, d] = timingLookup[str];
```
Is the @ts-ignore for this no longer needed? If so can you delete the comments related to it?

frank-weindel commented:
> @frank-weindel @wouterlucas I decided to try to update to Vite 5.x I hope that's alright :)
> 
> Crucial changes:
> 
> * Removed `variant` field since the new `FontFaceDescriptors` contract doesn't have it anymore. It seems to be compliant with https://www.w3.org/TR/css-fonts-3/#font-face-rule specification
> * Changed `tsconfig` module resolution to `Node16` in all variants due to [Cannot find module 'rollup/parseAst' or its corresponding type declarations. rollup/rollup#5199](https://github.com/rollup/rollup/issues/5199)
> 
> Let me know any of these require some additional checks to look for regression.

@M4tiz Excellent. Thanks! That's much appreciated.

Looks like there's an error in CI with this change related to the `examples` pnpm lock file:
```
Scope: all 3 workspace projects
 ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with examples/package.json

Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"

    Failure reason:
    specifiers in the lockfile ({"@lightningjs/renderer":"link:..","@stdlib/random-base-mt1[9](https://github.com/lightning-js/renderer/actions/runs/9614491706/job/26521477915?pr=300#step:7:10)937":"^0.2.1","@lightningjs/vite-plugin-import-chunk-url":"^0.3.0","vite":"^5.0.0"}) don't match specs in package.json ({"vite":"^5.0.0","@lightningjs/renderer":"link:.."})
```

frank-weindel commented on pnpm-lock.yaml:
```diff
@@ -88,256 +91,299 @@ importers:
       playwright:
         specifier: ^1.39.0
-        version: 1.39.0
+        version: 1.44.1
```
Something seems a bit off about how you're `pnpm` installing/upgrading the dependencies needed for this PR. I just did it myself on a [test branch](https://github.com/lightning-js/renderer/commit/3bc6886770493cb503366327ec68c8f1ff04bd6f) and I'm not seeing any version changes to any unrelated packages like this one in the `pnpm-lock.yaml` file. Are you using a mix of `npm`/`pnpm` by accident by any chance? The `package-lock.json` file should not be being updated.

It seems like the reason you're getting the visual regression failures is that your changes are updating `playwright` to this newer version which seems to change how things are rendering. I'd advise that you start the new dependency installation from scratch and avoid the unnecessary changes.

M4tiz replied:
You're right. It must've entered this state while I was tinkering with the dependencies. It's fixed now :)

frank-weindel replied:
Thanks. Looks good!
